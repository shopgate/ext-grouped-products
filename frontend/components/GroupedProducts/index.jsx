import React from 'react';
import PropTypes from 'prop-types';
import connect from './connector';
import styles from './style';

/**
 * The GroupedProducts component.
 * @param {props} param The component props.
 * @return {JSX}
 */
const GroupedProducts = ({ products }) => {
  if (!products.length) {
    return null;
  }

  return (
    <ul className={styles.container}>
      {products.map(({ name, id }) => (
        <li key={id}>{name}</li>
      ))}
    </ul>
  );
};

GroupedProducts.propTypes = {
  products: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default connect(GroupedProducts);
