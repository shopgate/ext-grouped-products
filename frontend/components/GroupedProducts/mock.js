import set from 'lodash/set';
import get from 'lodash/get';
import cloneDeep from 'lodash/cloneDeep';
import merge from 'lodash/merge';

const availability = {
  text: 'Lorem Ipsum',
  state: 'ok',
};

const stock = {
  ignoreQuantity: false,
  quantity: 30,
  info: 'Info',
  orderable: true,
  minOrderQuantity: 1,
  maxOrderQuantity: 1000,
};

const price = {
  currency: 'EUR',
  unitPrice: 5.0,
  unitPriceMin: 0,
  unitPriceStriked: 0,
  msrp: 0,
  discount: 0,
};

const product = {
  id: '1337',
  name: 'A shiny product',
  featuredImageUrl: 'https://cdn.acme.com/kitten.jpg',
  availability,
  price,
  stock,
};

let discounted = cloneDeep({
  ...product,
  id: '4711',
  name: 'A discounted product',
});

discounted = set(discounted, ['price', 'msrp'], 10.0);
discounted = set(discounted, ['price', 'discount'], 50);

let notOrderable = cloneDeep({
  ...product,
  id: '1234',
  name: 'A not orderable product',
});

notOrderable = set(notOrderable, ['stock', 'orderable'], false);
notOrderable = set(notOrderable, ['stock', 'quantity'], 0);

const mockedState = {
  product: {
    currentProduct: {
      productId: '1337',
    },
    productsById: {
      1337: {
        productData: product,
      },
    },
  },
};

export {
  mockedState,
  product as mockedProduct,
  discounted as mockedDiscountedProduct,
  notOrderable as mockedNotOrderableProduct,
};
