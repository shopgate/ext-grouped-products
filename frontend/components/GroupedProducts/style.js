import { css } from 'glamor';

const container = css({
  border: '2px solid red',
});

export default {
  container,
};
