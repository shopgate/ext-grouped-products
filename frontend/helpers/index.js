import { generateResultHash } from '@shopgate/pwa-common/helpers/redux';
import appConfig from '@shopgate/pwa-common/helpers/config';
import { SHOPGATE_CATALOG_GET_PRODUCT_CHILDREN, THEME_GMD } from '../constants';
import { maxQuantityPickerEntries as maxEntries } from '../config';

/**
 * Generates a hash to read and write with the productsByHash store.
 * @param {string} productId A product id.
 * @return {string}
 */
export const generateHash = productId => generateResultHash({
  pipeline: SHOPGATE_CATALOG_GET_PRODUCT_CHILDREN,
  productId,
});

/**
 * Creates an item list for the AddToCartPicker.
 * @param {Object} stock The stock property of a product.
 * @return {Array}
 */
export const createPickerItems = (stock) => {
  if (!stock) {
    return [];
  }

  const {
    ignoreQuantity,
    quantity,
  } = stock;

  const items = [];

  for (let i = 0; i <= maxEntries; i += 1) {
    if (ignoreQuantity) {
      if (items.length === maxEntries) {
        // When the quantity can be ignored the maximum number of entries can be created.
        break;
      }
    } else if (items.length === maxEntries || i >= quantity) {
      // When the quantity can't be ignored the maxium number of entries is limited.
      break;
    }

    const value = i + 1;

    items.push({
      label: `${value}`,
      value,
    });
  }

  return items;
};

/**
 * Checks if the current theme is the GMD theme.
 * @return {boolean}
 */
export const isGmdTheme = () => process.env.THEME.includes(THEME_GMD);

/**
 * Tells if the extension shall render flat buttons.
 * @return {boolean}
 */
export const renderFlatButtons = () => !isGmdTheme();

/**
 * Check if the favorite list feature is activated.
 * @return {boolean}
 */
export const hasFavorites = () => appConfig.hasFavorites;
