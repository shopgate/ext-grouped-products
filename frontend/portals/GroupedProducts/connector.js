import { connect } from 'react-redux';
import { getGroupedProducts } from '../../selectors';

/**
 * Maps state to props.
 * @param {Object} state State.
 * @returns {Object}
 */
const mapStateToProps = state => ({
  products: getGroupedProducts(state),
});

export default connect(mapStateToProps);
