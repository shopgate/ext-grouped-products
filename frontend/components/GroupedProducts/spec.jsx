import React from 'react';
import { shallow } from 'enzyme';
import { hasFavorites, renderFlatButtons } from '../../helpers/index';
import GroupedProducts from './index';

jest.mock('../../helpers/index', () => ({
  renderFlatButtons: jest.fn(),
  hasFavorites: jest.fn(),
}));

const withProducts = [
  { id: '1234' },
  { id: '1337' },
];
const withoutProducts = [];

describe('<GroupedProducts />', () => {
  it('should not render if there is no products', () => {
    const wrapper = shallow(<GroupedProducts
      products={withoutProducts}
      renderFlatButtons={renderFlatButtons}
      hasFavorites={hasFavorites}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').exists()).toBeFalsy();
  });

  it('should render if there are products', () => {
    const wrapper = shallow(<GroupedProducts
      products={withProducts}
      renderFlatButtons={renderFlatButtons}
      hasFavorites={hasFavorites}
    />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('List').exists()).toBeTruthy();
  });
});
