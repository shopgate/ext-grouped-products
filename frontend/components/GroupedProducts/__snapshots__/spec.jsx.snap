// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GroupedProducts /> should not render if there is no products 1`] = `""`;

exports[`<GroupedProducts /> should render if there are products 1`] = `
<List
  className="css-5bomak"
>
  <ListItem
    className="css-izecxm"
    isSelected={false}
    key="1234"
  >
    <Connect(Product)
      product={
        Object {
          "id": "1234",
        }
      }
    />
  </ListItem>
  <ListItem
    className="css-izecxm"
    isSelected={false}
    key="1337"
  >
    <Connect(Product)
      product={
        Object {
          "id": "1337",
        }
      }
    />
  </ListItem>
</List>
`;
